src/main/src/main/java/org/geoserver/feature/RetypingFeatureCollection.java

```diff
@@ -56,6 +59,13 @@ protected boolean canDelegate(FeatureVisitor visitor) {
         return ReTypingFeatureCollection.isTypeCompatible(visitor, target);
     }
 
+    @Override
+    public SimpleFeatureCollection subCollection(Filter filter) {
+        
+        SimpleFeatureCollection delegateCollection = delegate.subCollection(filter);
+        
+        return new ReTypingFeatureCollection(delegateCollection, target);
```

Comment:
This gave me pause for a moment, but retype can really only shave off attributes, so ti should be ok I guess.